### PR TITLE
build: disable BIP70 support by default

### PR DESCRIPTION
**Pull request by fanquake**
Disable BIP70 support in the GUI by default for `0.19.0` (for eventual removal in `0.20.0`?). 

Users who want to compile with BIP70 support enabled can pass `--enable-bip70` to `./configure`.

I've inverted the current `--disable-bip70` test to instead pass `--enable-bip70`.

Tested configurations on `macOS` (`protobuf` installed with `brew`).
Protobuf available and `./configure`:
```
Options used to compile and link:
  with wallet   = yes
  with gui / qt = yes
    with bip70  = no
```

Protobuf available and `./configure --enable-bip70`:
```
Options used to compile and link:
  with wallet   = yes
  with gui / qt = yes
    with bip70  = yes
```

Protobuf not available (i.e `brew unlink protobuf`) and `./configure`:
```
Options used to compile and link:
  with wallet   = yes
  with gui / qt = yes
    with bip70  = no
```

Protobuf not available and `./configure --enable-bip70`:
```
checking whether to build test_bitcoin-qt... yes
checking whether to build BIP70 support... configure: error: protobuf missing
```

TODO:
- [x] Remove `protobuf` from other Travis builds
- [ ] Documentation updates (mention that `protobuf` is now optional)?
- [ ] Could split release notes into GUI and build

### CODE REVIEW

**Comment by jonasschnelli** [comment]
Concept ACK
- IMO travis (not all jobs though) should build with BIP70 as long as the feature is available.
- Could we remove protobuf from the depends-build in the same pull?

**Comment by Sjors** [comment]
So is this removing BIP70 from the release as well? I'm reluctant to turn it off for developers by default while still shipping it. At minimum Travis should continue to check it in that case. But even then, GUI test coverage is too low to rely on that.

**Comment by jonasschnelli** [comment]
> So is this removing BIP70 from the release as well? 

Not for the release but for the gitian built binaries. We could re-add the BIP70 support to the gitian build. But as this is, it will disable BIP70 in gitian.

**Comment by jameshilliard** [comment]
I would recommend holding off on disabling support for BIP70 for release binaries until all major merchant processors have stopped requiring it. Dropping BIP70 support before the merchant payment processors stop requiring it would incentivize users to switch to less secure wallets that have BIP70 implementations with known unpatched vulnerabilities.

I would instead recommend making the deprecation warning more aggressive, maybe add to the deprecation warning that it's recommended users contact/inform the merchant that they are using an deprecated and insecure protocol. 

Unfortunately the largest merchant payment processor that requires BIP70 is still publishing misinformation in regards to the security of BIP70(such as disproven claims that BIP70 improves protection against MITM attacks) so it may be necessary in the deprecation warning to indicate that merchants should not be trusted to provide accurate information about BIP70 and that users should not follow any merchant recommendations to switch wallets.

**Comment by luke-jr** [comment]
@jameshilliard Do any merchants require BIP70 at this point? (A particular payment processor claims it does, but AFAIK their implementation is broken and already doesn't work with Core.)

**Comment by jameshilliard** [comment]
> Do any merchants require BIP70 at this point? (A particular payment processor claims it does, but AFAIK their implementation is broken and already doesn't work with Core.)

Well Core can make payments to that particular processor, the incompatible/non-standard part of their implementation introduces an unfixed security vulnerability but it doesn't seem to affect the ability to send payments(due to the incompatible/non-standard part being after the payment is broadcast by Core to the network). 

My worry is that prematurely removing BIP70 in Core would result in users switching to vulnerable wallets when they see the error message indicating they can't make a payment. 

Having a warning message encourage users to complain to the merchant about using a deprecated/insecure protocol on the other hand would help increase their support costs for processing BIP70 transactions which is the main reason they refuse to fix their security vulnerability or support BIP21. The idea is essentially to have the warning make their support costs for BIP21 lower than their BIP70 support costs by increasing merchant/customer complaints and support costs for BIP70.

**Comment by luke-jr** [comment]
Hate to say it, but we should probably do a release with something like #15064 before disabling BIP70 by default... :/

**Comment by laanwj** [comment]
ACK for 0.19.0

**Comment by TheBlueMatt** [comment]
Concept ACK

**Comment by achow101** [comment]
ACK c53da7c5f166a95f3ae37f817f8cd7b94812077a

Tested that the default configure does not download BIP 70 requests.

**Comment by dongcarl** [comment]
We probably should also remove `native_protobuf` and `protobuf` as default depends packages, with a flag to enable them explicitly.

**Comment by fanquake** [comment]
@dongcarl sure. I can also add/update documentation about Protobuf being optional.

**Comment by luke-jr** [comment]
@achow101 Since you have the sole ack here: did you consider the effect of essentially losing wallet metadata?

Maybe BIP70 support should be split between network/new transactions, and parsing/old transactions, and the latter retained in some form longer until we can migrate the metadata...

**Comment by achow101** [comment]
> did you consider the effect of essentially losing wallet metadata?

What metadata? AFAIK no BIP70 specific things are written to wallet files.

**Comment by luke-jr** [comment]
That's incorrect. There is BIP70-specific protobuf-encoded metadata written to wallets for each BIP70 transaction made.

File: configure.ac [hunk 1]
```diff
@@ -224,9 +224,9 @@ AC_ARG_ENABLE([zmq],
   [use_zmq=yes])
 AC_ARG_ENABLE([bip70],
   [AS_HELP_STRING([--disable-bip70],
```
Comment (ryanofsky) [inline configure.ac]: Should probably change the help string to --enable-bip70 so it is more comprehensible.

**Review by ryanofsky — APPROVED** [review]
I'm confused by these comments:

> > So is this removing BIP70 from the release as well?
> 
> Not for the release but for the gitian built binaries.

What's the difference between the "release" and the gitian built binaries? The gitian binaries along with the source tarball (where BIP70 will be disabled by default) *are* the release right?

But utACK a3fb4565348f14f19749aba9b74f4e5918aae4c3 assuming this does turn off bip70 by default in all release artifacts.

**Comment by practicalswift** [comment]
ACK e09913f1c47e693b0c6fafef55b9ca78e5f3abc0 -- diff looks correct

Smaller attack surface is better!

**Comment by shesek** [comment]
@gmaxwell written a [thoughtful comment](https://www.reddit.com/r/Bitcoin/comments/d87h71/why_dont_i_get_to_choose_bip_70_removed_by_default/f1cedy9/) on reddit explaining some of the motives behind this decision and the issues surrounding BIP 70. I thought it would be useful to copy this here for future reference:

> BIP70 has resulted in remotely exploitable vulnerabilities that would allow an attacker to steal the wallet's private keys.
> 
> BIP70 has had multiple serious privacy vulnerabilities, and the design of BIP70 where the client makes a connection to a requester specified host is generally prone to additional privacy vulnerabilities.
> 
> BIP70 requires access to a relatively complete x509 implementation as well as a TLS implementation, tens to hundreds of thousands of lines of security critical code that get exposed to the network.
> 
> BIP70 support has been blocking the removal of openssl as a dependency of the Bitcoin software. OpenSSL has a half dozen times or so been the origin of needing to perform short notice emergency upgrades of the software. Emergency upgrades deprive users of their own free choice of software and are an attack vector for introducing vulnerabilities. Additional in some cases OpenSSL's updates have come in the form of an opaque monolithic path that was hundreds of thousands of lines long, making actual review of their changes effectively impossible.
> 
> Bitpay's original incompatible modified version of BIP70, which was never used in Bitcoin Core but would have been needed for bitpay compatibility, introduced a serious vulnerability which which would allow bitpay (or someone who compromised their systems) to steal users coins or also just accidentally and almost invisibly take them via an easy mistake.
> 
> BIP70's implementation in bitcoin is extensively tied into the GUI and so uniform functionality cannot be provided for the RPC or CLI without essentially rewriting it, an effort that would not be justified given that the support is, as far as anyone can tell, largely unused. The lack of an automatable interface to it also makes it very difficult to subject to automated testing. The GUI tie-in also makes it difficult to separate BIP70 into another process in order to reduce the risk from vulnerabilities in the bip70 code. BIP70 also creates a dependency on protobuf, another piece of large network exposed code that could result in vulnerabilities or incompatibilities.
> 
> So, in summary: BIP70 in Core is largely unused, the implementation in Core has never been compatible with the incompatible version used by the only well known user (a company started using it after a decision had been made to remove it... which delayed the removal to see what would happen). BIP70's design is not particularly fit-for-purpose, it doesn't accomplish what people wanted it to do (thus the incompatible proprietary modifications and the lack of adoption), and the implementation has been a repeated source of privacy and security vulnerabilities. Attempts to improve BIP70 have been the source of security vulnerabilities. BIP70's implementation also requires pulling in an enormous amount of third party cryptographic code, which again, have frequently been the source of security vulnerabilities and emergency updates for the software.
> 
> And sure, lots of people justifiably dislike Bitpay-- but the removal of BIP70 started before bitpay was even using it and has essentially nothing to do with them beyond the fact that their usage, almost alone, of it isn't perceived to be enough justification to take the ongoing cost and risk of continuing to support it.
> 
> Of course, if you want to continue to personally use and support it-- you're free to do so. But it isn't reasonable to demand that other parties do so at their own cost, particularly without a clear justification as to how it benefits their users.

**Pull request merged**